File: 110.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution(object):
    def isBalanced(self, root):
        """
        :type root: TreeNode
        :rtype: bool
        """
        def helper(root):
            if not root: return 0
            right = helper(root.right)
            left = helper(root.left)
            if right == -1 or left == -1 or abs(right - left) > 1:
                return -1
            return 1 + max(right,left)
        return helper(root) != -1

File: 110.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    bool isBalanced(TreeNode* root) {
        return helper(root);
    }
    
    bool helper(TreeNode* root){
        if (root != NULL){
            int leftHeight = getDepth(root->left, 1);
            int rightHeight = getDepth(root->right, 1);
            if (abs(leftHeight - rightHeight) > 1){
                return false;
            }
            return helper(root->left) and helper(root->right);
        }
        return true;
    }
    int getDepth(TreeNode* root,int depth){
        if (root != NULL){
            return max(getDepth(root->left, depth + 1), getDepth(root->right, depth + 1));
        }
        return depth - 1;
    }
};
